### PR TITLE
Add Video Mask2Former

### PR DESCRIPTION
**Pull request by shivalikasingh95**
###  What does this PR do?

This PR adds Video Mask2Former model. 

Original repo: https://github.com/facebookresearch/Mask2Former/
Mask2Former for Video Instance Segmentation Paper: https://arxiv.org/abs/2112.10764

Co-authored with: @alaradirik 

- [x] Update model checkpoints
- [x] Update model cards
- [ ]  transfer model checkpoints to facebook organization 

### Who can review?
@alaradirik

### CODE REVIEW

File: tests/models/mask2former/test_modeling_mask2former.py [hunk 1]
```diff
@@ -317,20 +327,30 @@ def prepare_img():
     return image
 
 
+# We will verify our results on a video of cars
+def prepare_video():
```
Comment (alaradirik) [inline tests/models/mask2former/test_modeling_mask2former.py:331]: Just checking in with @amyeroberts to see if we should avoid using torchvision over here. For example, VideoMAE [tests](https://github.com/huggingface/transformers/blob/main/tests/models/videomae/test_modeling_videomae.py#LL350C4-L350C4) load a video file converted to a numpy array.

**Review by alaradirik — APPROVED** [review]
Left a few minor comments (some of them are my typos) but everything looks good to me!

Thank you for adding this, I believe this is our first video segmentation model :)

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model @shivalikasingh95! 

General structure is looking good and very nice and tidy code ❤️  

I suggest that instead of adapting the logic within the current `Mask2Former` layers and image processor, a new model `Mask2FormerVideo` is added, with its own image processor and modeling file. This would enable simpler logic in the forward passes for the respective models, and an image processor which can take and return videos directly. 

This PR is in a good state, so it should be fairly simple to make this change but of course let us know if you have any questions about this implementation.

**Comment by shivalikasingh95** [comment]
> I suggest that instead of adapting the logic within the current `Mask2Former` layers and image processor, a new model `Mask2FormerVideo` is added, with its own image processor and modeling file. This would enable simpler logic in the forward passes for the respective models, and an image processor which can take and return videos directly.
> 
> This PR is in a good state, so it should be fairly simple to make this change but of course let us know if you have any questions about this implementation.

Hey @amyeroberts , thanks a lot for the review :)

Regarding the structure...Alara and I actually had a discussion regarding this earlier...whether to add this as a separate model or not but we felt that since video mask2former is not very different from original mask2former, it would be alright if we just modify the existing implementation to handle both video and image. 

But I'd be happy to convert this into a separate model.
Just want to quickly double check with @alaradirik too before proceeding as this relates to our earlier discussion.

**Comment by alaradirik** [comment]
@shivalikasingh95 I'm fine with going either way since the video segmentation model only has minor differences but perhaps we could keep most of changes to the existing sub classes and add a new head class - `Mask2FormerForVideoSegmentation` and add it to the model docs. I think this would boost the visibility and usage of the model as well, what do you think?

Just asking for future models @amyeroberts, would we have a video_processing_xxx.py file and `VideoProcessorXXX` class to process videos? We have talked about creating video processing utilities with @shivalikasingh95 before but I wasn't sure about the best way to handle it.

**Comment by shivalikasingh95** [comment]
> @shivalikasingh95 I'm fine with going either way since the video segmentation model only has minor differences but perhaps we could keep most of changes to the existing sub classes and add a new head class - `Mask2FormerForVideoSegmentation` and add it to the model docs. I think this would boost the visibility and usage of the model as well, what do you think?

@alaradirik I think adding a new head class - `Mask2FormerForVideoSegmentation` is a really good idea! 

> Just asking for future models @amyeroberts, would we have a video_processing_xxx.py file and `VideoProcessorXXX` class to process videos? 

If we can add something like `Mask2FormerVideoProcessor` which can handle videos directly then that would be perfect.
______________________________________________________________________________________________

I'm not fully sure if it would make sense to add a separate modeling file altogether for video-mask2former since the authors wanted to show how easily we can use mask2former for video segmentation too.

Quote from the paper -:
_"We find Mask2Former also achieves state-of-the-art performance on video instance segmentation without modifying the architecture, the loss or even the training pipeline"_

And hence, implementation wise there isn't really much difference.
______________________________________________________________________________________________

I think adding a new head and video processor class should help in making the implementation cleaner.

But again, I'm not sure if a model can have both an ImageProcessor and VideoProcessor class. If you guys feel, this may not be the best approach then may be we can go for turning this into a separate model.

**Comment by amyeroberts** [comment]
> I'm not fully sure if it would make sense to add a separate modeling file altogether for video-mask2former 

I agree that having a separate model implementation isn't in line with the spirit of the model. However, at the moment, supporting video inputs does require a modification of architecture, as shown by the need for the `is_video` flag throughout `modeling_mask2formers.py`. 

> If we can add something like Mask2FormerVideoProcessor which can handle videos directly then that would be perfect.

I don't think a `VideoProcessor` class is necessary at the moment. We already have models with video inputs e.g. [VideoMAE](https://huggingface.co/docs/transformers/main/en/model_doc/videomae) which use image processors.
If we want to use the same processing class as the image model, adding a method such as `preprocess_video` is a possibility. This does mean the processor won't be compatible with the usual API, i.e. it could not be directly called `image_processor(video_inputs)`. However, the current method `post_process_video_instance_segmentation` also breaks this. Having a separate modeling file resolves the issue of one class handling both images and videos.

**Comment by shivalikasingh95** [comment]
> Having a separate modeling file resolves the issue of one class handling both images and videos.

Sure @amyeroberts. I understand now why adding a separate modeling file would make more sense. I'll take care of making this change.

>I don't think a VideoProcessor class is necessary at the moment. We already have models with video inputs e.g. [VideoMAE](https://huggingface.co/docs/transformers/main/en/model_doc/videomae) which use image processors.

Again makes sense, @amyeroberts. I only meant to suggest adding a VideoProcessor class for Mask2Former in case you guys were planning on introducing VideoProcessor classes in general as part of transformers library.

>If we want to use the same processing class as the image model, adding a method such as preprocess_video is a possibility. This does mean the processor won't be compatible with the usual API, i.e. it could not be directly called image_processor(video_inputs)

Would adding a separate image processor class, `VideoMask2FormerImageProcessor` make sense if we don't want to have the same image processing class for image and video models?
This way the usual API behaviour would not be broken. In this case, we can directly use `image_processor(video_inputs)`.

**Comment by amyeroberts** [comment]
>Would adding a separate image processor class, VideoMask2FormerImageProcessor make sense if we don't want to have the same image processing class for image and video models?
This way the usual API behaviour would not be broken. In this case, we can directly use image_processor(video_inputs).

@shivalikasingh95 - yep, I think that works!

**Comment by shivalikasingh95** [comment]
@alaradirik and @amyeroberts please feel free to review this PR.

I'm just getting a few failing CI checks due to [this error](https://app.circleci.com/pipelines/github/huggingface/transformers/62275/workflows/568ede06-a91a-4e5c-a02c-6bbafcbbcd64/jobs/766732). Would be great if I get some help on how to fix it.

File: src/transformers/models/mask2former/image_processing_mask2former.py [hunk 2]
```diff
@@ -1004,55 +1008,108 @@ def post_process_instance_segmentation(
...
+            )
 
             results.append({"segmentation": segmentation, "segments_info": segments})
+
         return results
 
+    def _post_process_instance_segmentation(
+        self,
+        num_predictions: int,
+        pred_masks: torch.Tensor,
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/image_processing_mask2former.py:1056]: The currently failing tests are coming from the typing here. All the torch objects in typing should be converted to strings. 
```suggestion
        pred_masks: "torch.Tensor",
```

Comment (shivalikasingh95) [reply src/transformers/models/mask2former/image_processing_mask2former.py:1056]: oh okay got it! Thanks for the help @amyeroberts :)

File: src/transformers/models/mask2former/convert_video_mask2former_checkpoint_to_pytorch.py [hunk 3]
```diff
@@ -1137,82 +1074,53 @@ def get_model_and_dataset_name(checkpoint_file: Path, is_video_ckpt: bool):
...
     config_dir: Path = args.configs_dir
     mask2former_dir: Path = args.mask2former_dir
-    is_video_ckpt = args.is_video_ckpt
+
     # append the path to the parents to mask2former dir
     sys.path.append(str(mask2former_dir.parent))
     sys.path.append(str(mask2former_dir))
 
-    # import original Mask2Former config and model from original source code repo
+    # import original VideoMask2Former model from original source code repo
```
Comment (alaradirik) [inline src/transformers/models/mask2former/convert_video_mask2former_checkpoint_to_pytorch.py:1088]: Just deleting extra whitespace
```suggestion
```

File: src/transformers/models/mask2former/image_processing_mask2former.py [hunk 4]
```diff
@@ -1004,55 +1008,108 @@ def post_process_instance_segmentation(
             scores = torch.nn.functional.softmax(mask_cls, dim=-1)[:, :-1]
             labels = torch.arange(num_classes, device=device).unsqueeze(0).repeat(num_queries, 1).flatten(0, 1)
 
-            scores_per_image, topk_indices = scores.flatten(0, 1).topk(num_queries, sorted=False)
+            # keep top-k predictions
```
Comment (alaradirik) [inline src/transformers/models/mask2former/image_processing_mask2former.py:1011]: Since Mask2Former Video's image processor is separated, could you undo the changes to `image_processing_mask2former.py`?

File: src/transformers/models/mask2former/image_processing_video_mask2former.py [hunk 5]
```diff
@@ -0,0 +1,1124 @@
...
+        return super().from_dict(image_processor_dict, **kwargs)
+
+    @property
+    def size_divisibility(self):
+        warnings.warn(
+            "The `size_divisibility` property is deprecated and will be removed in v4.27. Please use "
+            "`size_divisor` instead.",
+            FutureWarning,
+        )
+        return self.size_divisor
```
Comment (alaradirik) [inline src/transformers/models/mask2former/image_processing_video_mask2former.py:453]: Tagging @amyeroberts to double check this but we don't need to ensure backward compatibility over here since its a new image processor class.

```suggestion
```

File: src/transformers/models/mask2former/image_processing_video_mask2former.py [hunk 6]
```diff
@@ -0,0 +1,1124 @@
...
+
+    @property
+    def max_size(self):
+        warnings.warn(
+            "The `max_size` property is deprecated and will be removed in v4.27. Please use size['longest_edge']"
+            " instead.",
+            FutureWarning,
+        )
+        return self.size["longest_edge"]
```
Comment (alaradirik) [inline src/transformers/models/mask2former/image_processing_video_mask2former.py:462]: Same here, I think we can remove this.
```suggestion
```

File: docs/source/en/model_doc/mask2former.mdx [hunk 7]
```diff
@@ -30,7 +30,7 @@ The abstract from the video instance segmentation paper is the following:
 Tips:
 - Mask2Former uses the same preprocessing and postprocessing steps as [MaskFormer](maskformer). Use [`Mask2FormerImageProcessor`] or [`AutoImageProcessor`] to prepare images and optional targets for the model.
 - To get the final segmentation, depending on the task, you can call [`~Mask2FormerImageProcessor.post_process_semantic_segmentation`] or [`~Mask2FormerImageProcessor.post_process_instance_segmentation`] or [`~Mask2FormerImageProcessor.post_process_panoptic_segmentation`]. All three tasks can be solved using [`Mask2FormerForUniversalSegmentation`] output, panoptic segmentation accepts an optional `label_ids_to_fuse` argument to fuse instances of the target object/s (e.g. sky) together.
-- To get the final segmentation result for video instance segmentation task, you can call [`~Mask2FormerImageProcessor.post_process_video_instance_segmentation`]. This task can also be solved using output from [`Mask2FormerForUniversalSegmentation`].
+- To get the final segmentation result for video instance segmentation task, you can call [`~Mask2FormerImageProcessor.post_process_video_instance_segmentation`].
```
Comment (alaradirik) [inline docs/source/en/model_doc/mask2former.mdx:33]: Small typo over here:
```suggestion
- To get the final segmentation result for video instance segmentation task, you can call [`~Mask2FormerImageProcessor.post_process_instance_segmentation`].
```

File: src/transformers/models/mask2former/modeling_video_mask2former.py [hunk 8]
```diff
@@ -0,0 +1,2597 @@
...
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.mask_embedder = VideoMask2FormerMLPPredictionHead(self.hidden_size, self.hidden_size, mask_feature_size)
+
+    def forward(self, outputs: torch.Tensor, pixel_embeddings: torch.Tensor, attention_mask_target_size: int = None):
+        mask_embeddings = self.mask_embedder(outputs.transpose(0, 1))
+
+        outputs_mask = torch.einsum("bqc,btchw->bqthw", mask_embeddings, pixel_embeddings)
+        b, q, t, _, _ = outputs_mask.shape
+        # [B, Q, T, H, W] -> [B, Q, T*H*W] -> [B, h, Q, T*H*W] -> [B*h, Q, T*HW]
```
Comment (alaradirik) [inline src/transformers/models/mask2former/modeling_video_mask2former.py:2089]: Could you make the variable names and comments more self-descriptive?

**Review by alaradirik — APPROVED** [review]
I left a few minor comments but everything looks good to me. Video Mask2Former's image processor, model and test files are now separated and seem easy to use.

Thanks for adding this once again!

**Comment by alaradirik** [comment]
@amyeroberts could you do a final review, Video Mask2Former is a separate model now and the PR is in good shape :)

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work updating the model so far! 

I've left a lot of comments, however many of them are related and so there's less to change than it might first appear. 

In general, they can be grouped into some main areas: 
**Docstrings**
* Annotations for variables should be consistent, both in terms of style with other docstrings and with the rest of the modeling file. In some case, stated shapes didn't seem to align with the code. 
* The explanations should also be descriptive, explaining what the variable represents
* All arguments should be documented and be in the same order as the function signature

**Input / Output Types**
* I'm quite confused about what the input is to this model, in terms of shapes. In some parts of the code `num_frames` is referred to, in others it seems `pixel_values` is of shape `(batch_size, num_channels, height, width)`. Could you add details in docstrings and comments to both the image processor and model to clarify any places where this is ambiguous. 
* It seems that the input shape is different depending on whether the model is training or not. Is this correct? 

Model layers
* Model layers should (as much as possible) just take the `config` object to instantiate itself. 

**Tidying up code**
* Unused variables, attributes and methods should be removed. As should unused lines of code. 

**Tests**
* Some tests are skipped or overriden, but don't match the logic being checked in the rest of the library.

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work updating the model so far! 

I've left a lot of comments, however many of them are related and so there's less to change than it might first appear. 

In general, they can be grouped into some main areas: 
**Docstrings**
* Annotations for variables should be consistent, both in terms of style with other docstrings and with the rest of the modeling file. In some case, stated shapes didn't seem to align with the code. 
* The explanations should also be descriptive, explaining what the variable represents
* All arguments should be documented and be in the same order as the function signature

**Input / Output Types**
* I'm quite confused about what the input is to this model, in terms of shapes. In some parts of the code `num_frames` is referred to, in others it seems `pixel_values` is of shape `(batch_size, num_channels, height, width)`. Could you add details in docstrings and comments to both the image processor and model to clarify any places where this is ambiguous. 
* It seems that the input shape is different depending on whether the model is training or not. Is this correct? 

Model layers
* Model layers should (as much as possible) just take the `config` object to instantiate itself. 

**Tidying up code**
* Unused variables, attributes and methods should be removed. As should unused lines of code. 

**Tests**
* Some tests are skipped or overriden, but don't match the logic being checked in the rest of the library.

**Comment by Becomebright** [comment]
Does the latest version of transformers support Video Mask2Former?

File: src/transformers/models/mask2former/modeling_video_mask2former.py [hunk 9]
```diff
@@ -0,0 +1,2597 @@
...
+            transformer_decoder_hidden_states = outputs.transformer_decoder_hidden_states
+
+        output_auxiliary_logits = (
+            self.config.output_auxiliary_logits if output_auxiliary_logits is None else output_auxiliary_logits
+        )
+        if not output_auxiliary_logits:
+            auxiliary_logits = None
+
+        # [batch_size, num_queries, num_frames, height, width] -> [num_queries, num_frames, height, width]
+        mask_logits = mask_logits[0]
```
[the remark this answers is not in the document]
Comment (insomniaaac) [reply src/transformers/models/mask2former/modeling_video_mask2former.py:2576]: found code crash when batchsize>1, maybe shape around here still confusing.
